### PR TITLE
Implement wrapper for rsmt2d.Tree

### PR DESCRIPTION
**Pull request by evan-forbes**
This PR creates a wrapper for `NamespacedMerkleTree` that conforms to the `rsmt2d.Tree` interface, which will enable [`RetrieveBlockData`](https://github.com/lazyledger/lazyledger-core/pull/232) to properly repair the extended data square. It will also help close [#172](https://github.com/lazyledger/lazyledger-core/issues/172). 

ref [#15](https://github.com/lazyledger/rsmt2d/issues/15)

closes: #24

### CODE REVIEW

File: wrapper_test.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+	"github.com/lazyledger/rsmt2d"
+)
+
+const (
+	shareSize           = 256
+	adjustedMessageSize = shareSize - DefaultNamespaceIDLen
+)
+
+func TestPushErasuredNamespacedMerkleTree(t *testing.T) {
```
Comment (evan-forbes) [inline wrapper_test.go:14]: I wasn't sure what else I could test, so any suggestions are greatly appreciated.

File: go.mod [hunk 2]
```diff
@@ -5,4 +5,5 @@ go 1.14
 require (
 	github.com/google/gofuzz v1.2.0
 	github.com/lazyledger/merkletree v0.0.0-20201214195110-6901c4c3c75f
+	github.com/lazyledger/rsmt2d v0.0.0-20210121155026-23bf69f65287
```
Comment (liamsi) [inline go.mod:8]: IMO, the NMT should not need to know about rsmt2d. That is why I think the wrapper should live in ll-core. There all the dependencies need to come together anyways. And introducing the dependency here might cause some headaches in the future.

File: wrapper.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
+package nmt
+
+import (
+	"crypto/sha256"
+
+	"github.com/lazyledger/nmt/namespace"
+	"github.com/lazyledger/rsmt2d"
+)
+
+var paritySharesNamespaceID = namespace.ID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
```
Comment (liamsi) [inline wrapper.go:10]: Note that the tree already knows about this internally: https://github.com/lazyledger/nmt/blob/e0a317aa2d0ff84fcf1478527e4c90e6a13245d7/hasher.go#L77

File: wrapper.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		panic("tree size exceeded")
+
+	// if the namespace is included in the data, use that ns
+	case w.pushCount+1 <= w.squareSize/2:
+		copy(nsID, data[:DefaultNamespaceIDLen])
+
+	// if the data is erasure data use the parity ns
+	default:
+		copy(nsID, paritySharesNamespaceID)
+	}
```
Comment (liamsi) [inline wrapper.go:61]: Couldn't the caller do this instead? They already know which shares are parity shares. Then they could simply prefix the parity shares with `namespace.ID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}`. It would simplify the wrapper a bit. 

On another note, I think we should change the method signature for `Push` from `Push(nid, data []byte)` (currently) to `Push(data []byte)` for the underlying nmt anyways. I think I've mentioned this in some convos but I never got around opening a small issue explaining the rationale for that 😬 Will do in a sec.

File: wrapper_test.go (hunk 1 above)
Comment (liamsi) [reply wrapper_test.go:14]: I think there is not much to test as the wrapper is really simple. And that is good. Well done!

The only thing I'd suggest is to also cover the non-happy paths in unit tests (e.g. the few edge cases where this can panic/error).

**Review by liamsi — COMMENTED** [review]
IMO, the same can be achieved without any changes to the NMT. Namely by adding the wrapper to ll-core. Otherwise, the PR looks good to me. I want to think further about the implications of introducing rsmt2d as a dependency here before approving.

File: go.mod (hunk 2 above)
Comment (adlerjohn) [reply go.mod:8]: So what's the plan here, is the logic in this PR going to be moved instead of lazyledger-core?

**Comment by evan-forbes** [comment]
I'm going to close this PR for now. As mentioned in the comments, it makes sense to keep any new logic  in lazyledger-core. I think we will still need a wrapper, because we need to pass a `rsmt2d.Tree` to `rsmt.RepairExtendedDataSquare` when recovering the data. That tree also needs to be aware of our hard coded parity namespace for the second half of the tree.

File: wrapper.go (hunk 4 above)
Comment (evan-forbes) [reply wrapper.go:61]: +1 for changing the signature

We will need the wrapper to be aware of the parity share, as it will have to be passed to `rsmt2d.RepairExtendedDataSquare`
